jumanji/environments/routing/rware/utils.py

```diff
@@ -0,0 +1,865 @@
+    agent_dirs = jax.random.choice(
+        direction_key, possible_directions, shape=(num_agents,)
+    )
+
+    # sample request queue
+    key, queue_key = jax.random.split(key)
+    shelf_request_queue = jax.random.choice(
+        queue_key,
+        shelf_ids,
+        shape=(request_queue_size,),
+        replace=False,
+    )
+    requested_ids = jnp.zeros(shelf_ids.shape)
+    requested_ids = requested_ids.at[shelf_request_queue].set(1)
+
+    # spawn agents and shelves
+    agents = jax.vmap(spawn_agent)(agent_coords, agent_dirs)
+    shelves = jax.vmap(spawn_shelf)(shelf_coordinates, requested_ids)
+    return key, agents, shelves, shelf_request_queue
+
+
+def place_entity_on_grid(
+    grid: chex.Array,
+    channel: chex.Array,
+    entities: Entity,
+    entity_id: chex.Array,
+) -> chex.Array:
+    """Places an entity (Agent/Shelf) on the grid based on its
+    (x, y) position defined once spawned.
+
+    Args:
+        grid: the warehouse floor grid array.
+        channel: the grid channel index, either agents or shelves.
+        entities: a pytree of Agent or Shelf type containing entity information.
+        entity_id: unique ID identifying a specific entity.
+
+    Returns:
+        the warehouse grid with the specific entity in its position.
+    """
+    entity = tree_slice(entities, entity_id)
+    x, y = entity.position.x, entity.position.y
+    grid = grid.at[channel, x, y].set(entity_id + 1)
+    return grid
+
+
+def place_entities_on_grid(
+    grid: chex.Array, agents: Agent, shelves: Shelf
+) -> chex.Array:
+    """Place agents and shelves on the grid.
+
+    Args:
+        grid: the warehouse floor grid array.
+        agents: a pytree of Agent type containing agent information.
+        shelves: a pytree of Shelf type containing shelf information.
+
+    Returns:
+        the warehouse grid with all agents and shelves placed in their
+        positions.
+    """
+    agent_ids = get_entity_ids(agents)
+    shelf_ids = get_entity_ids(shelves)
+
+    # place agents and shelves on warehouse grid
+    def place_agents_scan(
+        grid_and_agents: Tuple[chex.Array, chex.Array], agent_id: chex.Array
+    ) -> Tuple[Tuple[chex.Array, chex.Array], None]:
+        grid, agents = grid_and_agents
+        grid = place_entity_on_grid(grid, _AGENTS, agents, agent_id)
+        return (grid, agents), None
+
+    def place_shelves_scan(
+        grid_and_shelves: Tuple[chex.Array, chex.Array], shelf_id: chex.Array
+    ) -> Tuple[Tuple[chex.Array, chex.Array], None]:
+        grid, shelves = grid_and_shelves
+        grid = place_entity_on_grid(grid, _SHELVES, shelves, shelf_id)
+        return (grid, shelves), None
+
+    (grid, _), _ = jax.lax.scan(place_agents_scan, (grid, agents), agent_ids)
+    (grid, _), _ = jax.lax.scan(place_shelves_scan, (grid, shelves), shelf_ids)
+    return grid
+
+
+def update_agent(
+    agents: Agent,
+    agent_id: chex.Array,
+    attr: str,
+    value: Union[chex.Array, Position],
```

Comment:
Would make reading the method easier if you replaced the last two args with `**agent_kwargs` although I'm sure how one goes about typing kwargs